src/session_state.rs

```diff
@@ -2,26 +2,45 @@
 ///
 ///
 use embedded_nal::IpAddr;
+use embedded_time::{duration::Seconds, Clock, Instant};
 use heapless::{consts, String, Vec};
 
-pub struct SessionState {
+pub struct SessionState<C>
+where
+    C: Clock,
+{
     pub connected: bool,
-    pub keep_alive_interval: u16,
+
+    /// (embedded_time doesn't support u16.)
+    pub keep_alive_interval: Seconds<u32>,
+
+    /// Timestamp of the last message sent to the broker, for keep-alive generation.
+    pub last_write_time: Option<Instant<C>>,
```

Comment:
I wouldn't allow a generic message to replace the ping message. Generic messages may use QoS = 0, which means fire-and-forget. Thus, if the broker is not responding, we may continue writing and never notice via the ping mechanism. Let's just ping at a fixed rate.